StandardServiceLocator aborts service discovery when Iterator.hasNext() throws LinkageError (e.g. NoClassDefFoundError) on OpenJDK 25

### Search first

- [x] I searched and no similar issues were found

### Description

`liquibase.servicelocator.StandardServiceLocator#findInstances` wraps only `iterator.next()` in a try/catch:

```java
  while (iterator.hasNext()) {          // <-- not in try/catch
      try {
          T next = iterator.next();
          ...
      } catch (Throwable e) {
          new ServiceLoadExceptionHandler().handleException(e);
      }
  }
```

The JDK's `ServiceLoader.LazyClassPathLookupIterator.hasNextService()` only catches `ServiceConfigurationError`. A `LinkageError` such as `NoClassDefFoundError` (raised when the JVM tries to resolve a missing class referenced in the provider's constructor signature, annotations, or superclass) escapes `hasNext()` directly and aborts the entire service-discovery loop. The exception handler never runs, and no providers get returned, even ones registered later in the same service file.

### Steps To Reproduce

Use any Liquibase deployment that has `liquibase-hibernate7` on the classpath but does not ship Spring (a perfectly valid configuration, Spring support is optional in the extension). `META-INF/services/liquibase.database.Database` from `liquibase-hibernate7` lists `HibernateSpringBeanDatabase` and `HibernateSpringPackageDatabase`, both of which transitively reference `org.springframework.core.io.support.ResourcePatternResolver`. When `DatabaseFactory.<init>` calls `findInstances(Database.class)`, you get:

```
Caused by: java.lang.NoClassDefFoundError: org/springframework/core/io/support/ResourcePatternResolver
    at java.base/java.lang.Class.getDeclaredConstructors0(Native Method)
    at java.base/java.lang.Class.privateGetDeclaredConstructors(Unknown Source)
    at java.base/java.lang.Class.getConstructor0(Unknown Source)
    at java.base/java.lang.Class.getConstructor(Unknown Source)
    at java.base/java.util.ServiceLoader.getConstructor(Unknown Source)
    at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNextService(Unknown Source)
    at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNext(Unknown Source)
    at java.base/java.util.ServiceLoader$1.hasNext(Unknown Source)
    at java.base/java.util.ServiceLoader$2.hasNext(Unknown Source)
    at liquibase.servicelocator.StandardServiceLocator.findInstances(StandardServiceLocator.java:24)
    at liquibase.database.DatabaseFactory.<init>(DatabaseFactory.java:39)
    ...
```

Notice the failure point: `StandardServiceLocator.java:24` is the `while (iterator.hasNext())` line. No `Database` providers are returned and bootstrap fails.

### Expected/Desired Behavior

A provider that fails to load (regardless of whether the JDK signals that via `ServiceConfigurationError`, `LinkageError`, or any other `Throwable`) should be logged and skipped, exactly as the existing comment-free intent of the try/catch around next() suggests. Subsequent providers in the same service file should still be discovered.


### Liquibase Version

5.0.2

### Database Vendor & Version

_No response_

### Liquibase Integration

Reproduced on OpenJDK 25.0.3-ea; the bug is JDK-agnostic in principle but is more easily triggered on modern JDKs where getConstructor is invoked during ServiceLoader.hasNextService.

### Liquibase Extensions

_No response_

### OS and/or Infrastructure Type/Provider

_No response_

### Additional Context

The JDK iterator's internal pending cursor is advanced before the failing class is resolved, so calling `hasNext()` again after a thrown `LinkageError` correctly moves on to the next provider, meaning the fix is local to `StandardServiceLocator` and does not require a JDK change.

## Suggested fix
  
Move the try/catch (`Throwable`) to surround the entire `hasNext()`/`next()` pair. Pull from the iterator unconditionally until it signals exhaustion, treating any error from either call as a skip-this-provider. The patch below is the implementation we are running in production as a higher-priority replacement; it can
drop straight into `StandardServiceLocator`:

```java
package liquibase.servicelocator;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Iterator;
import java.util.List;
import java.util.ServiceLoader;

import liquibase.Scope;
import liquibase.logging.Logger;

public class StandardServiceLocator implements ServiceLocator {
    @Override
    public int getPriority() {
        return 1;
    }

    @Override
    public <T> List<T> findInstances(Class<T> requiredInterface) {
        List<T> instances = new ArrayList<>();
        Logger log = Scope.getCurrentScope().getLog(getClass());
        ClassLoader classLoader = Scope.getCurrentScope().getClassLoader(true);
        Iterator<T> iterator = ServiceLoader.load(requiredInterface, classLoader).iterator();
        while (true) {
            try {
                if (!iterator.hasNext()) {
                    break;
                }
                T instance = iterator.next();
                log.fine("Loaded " + requiredInterface.getName()
                    + " instance " + instance.getClass().getName());
                instances.add(instance);
            } catch (Throwable e) {
                // LinkageError (e.g. NoClassDefFoundError) can escape JDK ServiceLoader.hasNext()
                // because the JDK only catches ServiceConfigurationError internally.
                new ServiceLoadExceptionHandler().handleException(e);
            }
        }
        return Collections.unmodifiableList(instances);
    }
}
```

The only structural change vs. the current code is that the while-condition is moved inside the try. Behavior is unchanged for the happy path and for `ServiceConfigurationError`-style failures (which the JDK already stashes and rethrows from next()).

## Workaround
  
Until this is fixed upstream, downstream projects can ship their own `ServiceLocator` with priority > 1 and the loop above.


### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR (Thank you!)

--- a/liquibase-standard/src/main/java/liquibase/servicelocator/StandardServiceLocator.java
+++ b/liquibase-standard/src/main/java/liquibase/servicelocator/StandardServiceLocator.java
@@ -21,8 +21,12 @@ public class StandardServiceLocator implements ServiceLocator {
 
         final Logger log = Scope.getCurrentScope().getLog(getClass());
         final Iterator<T> services = ServiceLoader.load(interfaceType, Scope.getCurrentScope().getClassLoader(true)).iterator();
-        while (services.hasNext()) {
+        // Guard hasNext() too: ServiceLoader may throw LinkageError (e.g. NoClassDefFoundError) from it.
+        while (true) {
             try {
+                if (!services.hasNext()) {
+                    break;
+                }
                 final T service = services.next();
                 log.fine("Loaded "+interfaceType.getName()+" instance "+service.getClass().getName());
                 allInstances.add(service);

--- a/liquibase-standard/src/test/groovy/liquibase/servicelocator/StandardServiceLocatorTest.groovy
+++ b/liquibase-standard/src/test/groovy/liquibase/servicelocator/StandardServiceLocatorTest.groovy
@@ -1,8 +1,8 @@
 package liquibase.servicelocator
 
+import liquibase.Scope
 import liquibase.change.Change
 import liquibase.changelog.ChangeLogHistoryService
-import liquibase.command.CommandStep
 import liquibase.database.Database
 import liquibase.database.DatabaseConnection
 import liquibase.datatype.LiquibaseDataType
@@ -23,9 +23,18 @@ import liquibase.structure.DatabaseObject
 import liquibase.util.TestUtil
 import org.junit.Assume
 import spock.lang.Specification
+import spock.lang.TempDir
 import spock.lang.Unroll
 
+import javax.tools.ToolProvider
+import java.nio.charset.StandardCharsets
+import java.nio.file.Files
+import java.nio.file.Path
+
 class StandardServiceLocatorTest extends Specification {
+
+    @TempDir
+    Path tempDir
 
     @Unroll("#featureName: #type.name")
     def "all classes are listed in service loader files"() {
@@ -67,5 +76,128 @@ class StandardServiceLocatorTest extends Specification {
                 Executor.class,
                 DatabaseConnection.class,
         ]
+    }
+
+    def "findInstances continues after NoClassDefFoundError from ServiceLoader.hasNext"() {
+        given:
+        def compiler = ToolProvider.getSystemJavaCompiler()
+        Assume.assumeNotNull("JDK JavaCompiler required to build broken provider fixtures", compiler)
+
+        def classesDir = tempDir.resolve("classes")
+        Files.createDirectories(classesDir)
+        compileServiceFixtures(classesDir)
+        Files.delete(classesDir.resolve("servicelocator/testfixture/MissingDependency.class"))
+
+        URLClassLoader serviceClassLoader = new URLClassLoader(
+                [classesDir.toUri().toURL()] as URL[],
+                getClass().getClassLoader()
+        )
+        def serviceType = serviceClassLoader.loadClass("servicelocator.testfixture.OptionalDependencyService")
+
+        when:
+        def instances = Scope.child([(Scope.Attr.classLoader.name()): serviceClassLoader], {
+            return new StandardServiceLocator().findInstances(serviceType)
+        } as Scope.ScopedRunnerWithReturn)
+
+        then:
+        instances.size() == 2
+        instances*.getClass().name as Set == [
+                "servicelocator.testfixture.GoodProviderBefore",
+                "servicelocator.testfixture.GoodProviderAfter",
+        ] as Set
+
+        cleanup:
+        serviceClassLoader?.close()
+    }
+
+    def "findInstances returns empty list when every provider fails to load"() {
+        given:
+        def compiler = ToolProvider.getSystemJavaCompiler()
+        Assume.assumeNotNull("JDK JavaCompiler required to build broken provider fixtures", compiler)
+
+        def classesDir = tempDir.resolve("classes-all-bad")
+        Files.createDirectories(classesDir)
+        compileServiceFixtures(classesDir)
+        Files.delete(classesDir.resolve("servicelocator/testfixture/MissingDependency.class"))
+
+        def servicesFile = classesDir.resolve("META-INF/services/servicelocator.testfixture.OptionalDependencyService")
+        Files.write(servicesFile, "servicelocator.testfixture.BadProvider\n".getBytes(StandardCharsets.UTF_8))
+
+        URLClassLoader serviceClassLoader = new URLClassLoader(
+                [classesDir.toUri().toURL()] as URL[],
+                getClass().getClassLoader()
+        )
+        def serviceType = serviceClassLoader.loadClass("servicelocator.testfixture.OptionalDependencyService")
+
+        when:
+        def instances = Scope.child([(Scope.Attr.classLoader.name()): serviceClassLoader], {
+            return new StandardServiceLocator().findInstances(serviceType)
+        } as Scope.ScopedRunnerWithReturn)
+
+        then:
+        instances.isEmpty()
+
+        cleanup:
+        serviceClassLoader?.close()
+    }
+
+    private void compileServiceFixtures(Path classesDir) {
+        def srcDir = tempDir.resolve("src-" + classesDir.fileName)
+        def pkgDir = srcDir.resolve("servicelocator/testfixture")
+        Files.createDirectories(pkgDir)
+
+        writeJava(pkgDir.resolve("OptionalDependencyService.java"), """
+            package servicelocator.testfixture;
+            public interface OptionalDependencyService {
+            }
+            """)
+        writeJava(pkgDir.resolve("MissingDependency.java"), """
+            package servicelocator.testfixture;
+            public class MissingDependency {
+            }
+            """)
+        writeJava(pkgDir.resolve("GoodProviderBefore.java"), """
+            package servicelocator.testfixture;
+            public class GoodProviderBefore implements OptionalDependencyService {
+                public GoodProviderBefore() {}
+            }
+            """)
+        writeJava(pkgDir.resolve("BadProvider.java"), """
+            package servicelocator.testfixture;
+            public class BadProvider extends MissingDependency implements OptionalDependencyService {
+                public BadProvider() {}
+            }
+            """)
+        writeJava(pkgDir.resolve("GoodProviderAfter.java"), """
+            package servicelocator.testfixture;
+            public class GoodProviderAfter implements OptionalDependencyService {
+                public GoodProviderAfter() {}
+            }
+            """)
+
+        def sources = Files.walk(srcDir)
+                .filter { Files.isRegularFile(it) && it.toString().endsWith(".java") }
+                .map { it.toFile() }
+                .toList()
+
+        def compiler = ToolProvider.getSystemJavaCompiler()
+        def fileManager = compiler.getStandardFileManager(null, null, StandardCharsets.UTF_8)
+        def compilationUnits = fileManager.getJavaFileObjectsFromFiles(sources)
+        def options = ["-d", classesDir.toAbsolutePath().toString()]
+        def task = compiler.getTask(null, fileManager, null, options, null, compilationUnits)
+        assert task.call() : "Failed to compile service locator test fixtures"
+        fileManager.close()
+
+        def servicesDir = classesDir.resolve("META-INF/services")
+        Files.createDirectories(servicesDir)
+        Files.write(servicesDir.resolve("servicelocator.testfixture.OptionalDependencyService"), """
+            servicelocator.testfixture.GoodProviderBefore
+            servicelocator.testfixture.BadProvider
+            servicelocator.testfixture.GoodProviderAfter
+            """.stripIndent().trim().getBytes(StandardCharsets.UTF_8))
+    }
+
+    private static void writeJava(Path path, String source) {
+        Files.write(path, source.stripIndent().trim().getBytes(StandardCharsets.UTF_8))
     }
 }

